rpm/0005-Retain-symlink-chains.patch

```diff
@@ -0,0 +1,209 @@
+it was when the application was started.
+
+Similar problems occur with symlinks that go though dynamic/virtual
+entries such as /proc/self.
+
+Retain symlink chains by duplicating symlinks as-is instead of resolving the
+final target.
+
+Remove the now unnecessary special case code that was used to rectify
+/proc/PID/xxx type symlinks back into /proc/self/XXX form.
+
+Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>
+---
+ src/fcopy/main.c | 136 +++++++++++++++++++----------------------------
+ 1 file changed, 56 insertions(+), 80 deletions(-)
+
+diff --git a/src/fcopy/main.c b/src/fcopy/main.c
+index 31810de9..9fd922ed 100644
+--- a/src/fcopy/main.c
++++ b/src/fcopy/main.c
+@@ -22,6 +22,7 @@
+ #include <ftw.h>
+ #include <errno.h>
+ #include <pwd.h>
++#include <limits.h>
+ 
+ #include <fcntl.h>
+ #ifndef O_PATH
+@@ -180,77 +181,58 @@ static void mkdir_attr(const char *fname, mode_t mode, uid_t uid, gid_t gid) {
+ 	}
+ }
+ 
+-static char *proc_pid_to_self(const char *target) {
+-	assert(target);
+-	char *use_target = 0;
+-	char *proc_pid = 0;
+-
+-	if (!(use_target = realpath(target, NULL)))
+-		goto done;
+-
+-	// target is under /proc/<PID>?
+-	static const char proc[] = "/proc/";
+-	if (strncmp(use_target, proc, sizeof(proc) - 1))
+-		goto done;
+-
+-	int digit = use_target[sizeof(proc) - 1];
+-	if (digit < '1' || digit > '9')
+-		goto done;
+-
+-	// check where /proc/self points to
+-	static const char proc_self[] = "/proc/self";
+-	if (!(proc_pid = realpath(proc_self, NULL)))
+-		goto done;
+-
+-	// redirect /proc/PID/xxx -> /proc/self/XXX
+-	size_t pfix = strlen(proc_pid);
+-	if (strncmp(use_target, proc_pid, pfix))
+-		goto done;
+-
+-	if (use_target[pfix] != 0 && use_target[pfix] != '/')
+-		goto done;
+-
+-	char *tmp;
+-	if (asprintf(&tmp, "%s%s", proc_self, use_target + pfix) != -1) {
+-		if (arg_debug)
+-			fprintf(stderr, "SYMLINK %s\n  -->   %s\n", use_target, tmp);
+-		free(use_target);
+-		use_target = tmp;
+-	}
+-	else
+-		errExit("asprintf");
+-
+-done:
+-	if (proc_pid)
+-		free(proc_pid);
+-	return use_target;
+-}
+-
+-void copy_link(const char *target, const char *linkpath, mode_t mode, uid_t uid, gid_t gid) {
+-	(void) mode;
+-	(void) uid;
+-	(void) gid;
++void copy_link(const char *target, const char *linkpath) {
++	int failed = 1;
++	char *linkdata = NULL;
+ 
+ 	// if the link is already there, don't create it
+ 	struct stat s;
+ 	if (lstat(linkpath, &s) == 0)
+-	       return;
+-
+-	char *rp = proc_pid_to_self(target);
+-	if (rp) {
+-		if (symlink(rp, linkpath) == -1) {
+-			free(rp);
+-			goto errout;
+-		}
+-		free(rp);
++		goto success;
++
++	// read source symlink
++	if (lstat(target, &s) == -1)
++		goto failure;
++
++	if (!S_ISLNK(s.st_mode))
++		goto failure;
++
++	ssize_t linksize = s.st_size ? (s.st_size + 1) : PATH_MAX;
++	if (!(linkdata = malloc(linksize)))
++	    goto failure;
++
++	ssize_t rc = readlink(target, linkdata, linksize);
++	if (rc < 0 ) {
```

Comment:
There is an extra space here.

Also I wonder what's up with the indentations, this looks like it's just one space but it might just be the patch messing it up. I guess I need to check it applied.